### PR TITLE
Add Modifications and RootFeature

### PR DESCRIPTION
**Pull request by s9latimm**
Add modifications to test broken FMs, refactor root nodes and clean up public interface.

- [x] Add removal of edges, children and features as modifications
  - [x] Enable creation of broken FM through inheritance
- [x] Remove `Feature::isRoot` to prevent misconception
  - [x] Add `RootFeature` class
  - [x] Refactor builder
  - [x] Refactor comparisons to root with `llvm::isa`
- [x] Clean up unused / deprecated constructors
- [x] Update python bindings
  - [x] Add `RootFeature`
  - [x] Add `feature_model.get_feature`
  - [x] Clean up modifications / deprecated methods
- [x] add `venv/` to gitignore

### CODE REVIEW

File: bindings/python/vara-feature/pybind_Feature.cpp [hunk 1]
```diff
@@ -80,13 +81,6 @@ void init_feature_module_binary_feature(py::module &M) {
       .def(py::init<std::string, bool>())
       .def(py::init<std::string, bool,
                     std::vector<vara::feature::FeatureSourceRange>>())
-      .def(py::init<std::string, bool,
-                    std::vector<vara::feature::FeatureSourceRange>,
-                    vara::feature::Feature *>())
-      .def(py::init<std::string, bool,
-                    std::vector<vara::feature::FeatureSourceRange>,
-                    vara::feature::Feature *,
-                    std::vector<vara::feature::FeatureTreeNode *>>())
```
Comment (s9latimm) [inline bindings/python/vara-feature/pybind_Feature.cpp:89]: @padupr do you need these constructors?

File: bindings/python/vara-feature/pybind_Feature.cpp [hunk 2]
```diff
@@ -45,6 +42,10 @@ void init_feature_module_feature(py::module &M) {
       .def("__hash__", &vf::Feature::hash)
       .def_property_readonly("name", &vf::Feature::getName,
                              R"pbdoc(The name of the feature.)pbdoc")
+      .def(
+          "is_root",
+          [](vf::Feature &F) { return llvm::isa<vf::RootFeature>(F); },
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_Feature.cpp]:
```suggestion
          [](const vf::Feature &F) { return llvm::isa<vf::RootFeature>(F); },
```
I, think this should be const, otherwise, we could get a problem when we call python functions on an underlying const object.

File: bindings/python/vara-feature/pybind_FeatureModel.cpp [hunk 3]
```diff
@@ -22,6 +22,12 @@ void init_feature_model_module(py::module &M) {
       .def("get_root", &vf::FeatureModel::getRoot,
            py::return_value_policy::reference,
            R"pbdoc(Returns the root Feature.)pbdoc")
+      .def(
+          "get_feature",
+          [](vf::FeatureModel &FM, const std::string &Name) {
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureModel.cpp:27]: Do we get a ref here from the python side? Not sure, normally, if we would call a member function on an C++ object this would be a pointer.

File: include/vara/Feature/Feature.h [hunk 4]
```diff
@@ -278,6 +257,21 @@ class NumericFeature : public Feature {
...
   ValuesVariantType Values;
 };
+
+//===----------------------------------------------------------------------===//
+//                          RootFeature Class
+//===----------------------------------------------------------------------===//
+
+class RootFeature : public Feature {
+public:
+  RootFeature(string Name = "root")
```
Comment (vulder) [inline include/vara/Feature/Feature.h]:
```suggestion
  explicit RootFeature(string Name = "root")
```
Ctors with only one parameter should be explicit to prevent unwanted implicit conversions.

File: include/vara/Feature/FeatureModel.h [hunk 5]
```diff
@@ -438,16 +446,17 @@ template <typename... Rules> class FeatureModelConsistencyChecker {
...
   }
 };
 
 struct CheckFeatureParentChildRelationShip {
   static bool check(FeatureModel &FM) {
     return std::all_of(FM.begin(), FM.end(), [](Feature *F) {
-      return F->isRoot() ||
-             // Every parent of a Feature needs to have the Feature as a child.
+      return llvm::isa<RootFeature>(F) ||
+             // Every parent of a Feature needs to have that as a child.
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h]:
```suggestion
             // Every parent of a Feature needs to have it as a child.
```

File: include/vara/Feature/FeatureModel.h [hunk 6]
```diff
@@ -438,16 +446,17 @@ template <typename... Rules> class FeatureModelConsistencyChecker {
 struct EveryFeatureRequiresParent {
   static bool check(FeatureModel &FM) {
     return std::all_of(FM.begin(), FM.end(), [](Feature *F) {
-      return F->isRoot() || (F->getParent() != nullptr);
+      return llvm::isa<RootFeature>(F) ||
+             (F->getParent() && F->getParentFeature());
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h]: Why do we need the extra `getParentFeature` check?

File: include/vara/Feature/FeatureModel.h [hunk 7]
```diff
@@ -164,15 +166,15 @@ class FeatureModelBuilder : private FeatureModel {
   template <typename FeatureTy, typename... Args,
             typename = typename std::enable_if_t<
                 std::is_base_of_v<Feature, FeatureTy>, int>>
-  Feature *makeFeature(const std::string &FeatureName, Args... FurtherArgs) {
+  FeatureTy *makeFeature(const std::string &FeatureName, Args... FurtherArgs) {
     if (!Features
              .try_emplace(FeatureName,
                           std::make_unique<FeatureTy>(
                               FeatureName, std::move(FurtherArgs)...))
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h]:
```suggestion
  FeatureTy *makeFeature(const std::string &FeatureName, Args &&...FurtherArgs) {
    if (!Features
             .try_emplace(FeatureName,
                          std::make_unique<FeatureTy>(
                              FeatureName, std::forward<Args>(FurtherArgs)...))
```

File: include/vara/Feature/FeatureModel.h [hunk 8]
```diff
@@ -164,15 +166,15 @@ class FeatureModelBuilder : private FeatureModel {
   template <typename FeatureTy, typename... Args,
             typename = typename std::enable_if_t<
                 std::is_base_of_v<Feature, FeatureTy>, int>>
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h]:
```suggestion
  template <typename FeatureTy, typename... Args,
            std::enable_if_t<std::is_base_of_v<Feature, FeatureTy>, int> = 0>
```

File: include/vara/Feature/OrderedFeatureVector.h [hunk 9]
```diff
@@ -22,7 +22,8 @@ class OrderedFeatureVector {
   OrderedFeatureVector(std::initializer_list<Feature *> Init) { insert(Init); }
   template <class FeatureIterTy>
   OrderedFeatureVector(FeatureIterTy Begin, FeatureIterTy End) {
```
Comment (vulder) [inline include/vara/Feature/OrderedFeatureVector.h]:
```suggestion
  OrderedFeatureVector(FeatureIterTy &&Begin, FeatureIterTy &&End) {
```

File: include/vara/Feature/OrderedFeatureVector.h [hunk 10]
```diff
@@ -42,7 +49,8 @@ class OrderedFeatureVector {
 
   template <class FeatureIterTy>
   void insert(FeatureIterTy Begin, FeatureIterTy End) {
```
Comment (vulder) [inline include/vara/Feature/OrderedFeatureVector.h]:
```suggestion
  void insert(FeatureIterTy &&Begin, FeatureIterTy &&End) {
```

File: include/vara/Feature/FeatureModel.h [hunk 11]
```diff
@@ -164,15 +166,15 @@ class FeatureModelBuilder : private FeatureModel {
...
-  Feature *makeFeature(const std::string &FeatureName, Args... FurtherArgs) {
+  FeatureTy *makeFeature(const std::string &FeatureName, Args... FurtherArgs) {
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h]:
```suggestion
  FeatureTy *makeFeature(std::string FeatureName, Args... FurtherArgs) {
```

File: unittests/Feature/BinaryFeature.cpp [hunk 12]
```diff
@@ -44,9 +42,7 @@ TEST(BinaryFeature, BinaryFeatureChildren) {
   EXPECT_EQ(
       std::distance(FM->getFeature("a")->begin(), FM->getFeature("a")->end()),
       1);
-  if (auto *F =
-          llvm::dyn_cast<vara::feature::Feature>(*FM->getFeature("a")->begin());
-      F) {
+  if (auto *F = llvm::dyn_cast<Feature>(*FM->getFeature("a")->begin()); F) {
```
Comment (vulder) [inline unittests/Feature/BinaryFeature.cpp]:
```suggestion
  if (auto *F = llvm::dyn_cast<Feature>(*FM->getFeature("a")->begin())) {
```
If you don't have a more complex check of need the variable defined in the else branch you can just use an normal if here without initializer.

File: unittests/Feature/FeatureModel.cpp [hunk 13]
```diff
@@ -45,20 +34,19 @@ TEST(FeatureModel, cloneUnique) {
...
   FM.reset();
 
-  // The inner EXPECT_TRUE just handles the nodiscard of isRoot
+  // The inner EXPECT_TRUE just handles the nodiscard of getParent
   // NOLINTNEXTLINE
-  ASSERT_EXIT(EXPECT_TRUE(FM->getFeature("a")->isRoot()),
-              testing::KilledBySignal(SIGSEGV), ".*");
-  EXPECT_TRUE(Clone->getFeature("a")->isRoot());
+  EXPECT_DEATH(EXPECT_TRUE(FM->getFeature("a")->getParent()), ".*");
+  EXPECT_FALSE(Clone->getFeature("a")->getParent());
```
Comment (vulder) [inline unittests/Feature/FeatureModel.cpp]: Shouldn't this become a `isa` check for RootNode instead of an empty parent check?

File: unittests/Feature/NumericFeature.cpp [hunk 14]
```diff
@@ -60,9 +58,7 @@ TEST(NumericFeature, NumericFeatureChildren) {
   EXPECT_EQ(
       std::distance(FM->getFeature("a")->begin(), FM->getFeature("a")->end()),
       1);
-  if (auto *F =
-          llvm::dyn_cast<vara::feature::Feature>(*FM->getFeature("a")->begin());
-      F) {
+  if (auto *F = llvm::dyn_cast<Feature>(*FM->getFeature("a")->begin()); F) {
```
Comment (vulder) [inline unittests/Feature/NumericFeature.cpp]:
```suggestion
  if (auto *F = llvm::dyn_cast<Feature>(*FM->getFeature("a")->begin())) {
```

File: unittests/Feature/RootFeature.cpp [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+  assert(FM);
+
+  EXPECT_EQ(
+      std::distance(FM->getFeature("a")->begin(), FM->getFeature("a")->end()),
+      1);
+  if (auto *F = llvm::dyn_cast<Feature>(*FM->getFeature("a")->begin()); F) {
+    EXPECT_EQ("aa", F->getName());
+  } else {
+    FAIL();
+  }
```
Comment (vulder) [inline unittests/Feature/RootFeature.cpp]:
```suggestion
  Feature *FirstChild = llvm::dyn_cast<Feature>(*FM->getFeature("a")->begin());
  EXPECT_NE(FirstChild, nullptr);
  EXPECT_EQ("aa", F->getName());
```
This produces more precise error messages.

**Review by vulder — CHANGES_REQUESTED** [review]

File: include/vara/Feature/FeatureModel.h (hunk 6 above)
Comment (s9latimm) [reply include/vara/Feature/FeatureModel.h]: In this loop we only iterate over feature nodes, so relationships parents are not covered. If we assert, that each feature has an ancestor, this implies that also every relationship has a parent of some kind (assuming we prohibit directly nested relationships).

File: unittests/Feature/FeatureModel.cpp (hunk 13 above)
Comment (s9latimm) [reply unittests/Feature/FeatureModel.cpp]: Indeed.

File: unittests/Feature/RootFeature.cpp (hunk 15 above)
Comment (s9latimm) [reply unittests/Feature/RootFeature.cpp]: I agree, but Clang-Tidy wants me to assure the pointer cannot be null in this block -- not even `ASSERT_` satisfies here.

File: include/vara/Feature/FeatureModel.h (hunk 7 above)
Comment (s9latimm) [reply include/vara/Feature/FeatureModel.h]: Moving `FeatureName` during the creation breaks, as it is needed for the emplacement afterwards.

File: bindings/python/vara-feature/pybind_FeatureModel.cpp (hunk 3 above)
Comment (s9latimm) [reply bindings/python/vara-feature/pybind_FeatureModel.cpp:27]: So far we always used refs as arguments and looking at pybind11s documentation, they are doing so as well -- idk.
It seems like ptrs work too, at least they do not produce errors yet.

Comment (vulder) [reply bindings/python/vara-feature/pybind_FeatureModel.cpp:27]: ok then let's stay with refs

File: include/vara/Feature/FeatureModel.h (hunk 6 above)
Comment (vulder) [reply include/vara/Feature/FeatureModel.h]: But how does checking if the parent is a Feature help if we don't iterate over the other nodes?

File: unittests/Feature/RootFeature.cpp (hunk 15 above)
Comment (vulder) [reply unittests/Feature/RootFeature.cpp]: Does a assert(FirstChild) help?

Comment (s9latimm) [reply unittests/Feature/RootFeature.cpp]: Yes, but `Release` will still fail.

Comment (vulder) [reply unittests/Feature/RootFeature.cpp]: ok, your right, then let's keep this.

File: include/vara/Feature/FeatureModel.h (hunk 6 above)
Comment (s9latimm) [reply include/vara/Feature/FeatureModel.h]: If we have a feature in a relationship, the parent node will only be of that kind. So if we find an parent feature (which can't be the actual parent), this means, that the relationship node itself has to have a parent, which again should be a feature. So we implicitly check if every relationship has a parent as well.

tl;dr: If we check that features are connected to a parent feature, this may be transitive over a relationship -- so the relationship itself has to have a parent in the first place.

File: include/vara/Feature/FeatureModel.h (hunk 7 above)
Comment (vulder) [reply include/vara/Feature/FeatureModel.h]: ok I remove the move

File: bindings/python/vara-feature/pybind_Feature.cpp (hunk 1 above)
Comment (vulder) [reply bindings/python/vara-feature/pybind_Feature.cpp:89]: nope, je does not need them

Comment (padupr) [reply bindings/python/vara-feature/pybind_Feature.cpp:89]: I don't think so.
